x-pack/plugins/security_solution/public/resolver/store/selectors.ts

```diff
@@ -122,3 +132,93 @@ function composeSelectors<OuterState, InnerState, ReturnValue>(
 ): (state: OuterState) => ReturnValue {
   return (state) => secondSelector(selector(state));
 }
+
+const spatiallyIndexedEntities = createSelector(
+  (state: ResolverState) => state,
+  function (state) {
+    const indexedProcessTree = indexedProcessTreeFactory(
+      dataSelectors.graphableProcesses(state.data)
+    );
+    const processNodeViewWidth = 360;
+    const processNodeViewHeight = 120;
+    const widths = dataSelectors.widthsOfProcessSubtrees(indexedProcessTree);
+
+    const positions = dataSelectors.processPositions(indexedProcessTree, widths);
+
+    const edgeLineSegments = dataSelectors.processEdgeLineSegments(
+      indexedProcessTree,
+      widths,
+      positions
+    );
+
+    const tree: rbush<IndexedEntity> = new rbush();
+    const processesToIndex: IndexedProcessNode[] = [];
+    const edgeLineSegmentsToIndex: IndexedEdgeLineSegment[] = [];
+    for (const [processEvent, position] of positions) {
+      const transformedPosition = applyMatrix3(position, dataSelectors.isometricTransformMatrix);
+      const [nodeX, nodeY] = transformedPosition;
+      const indexedEvent: IndexedProcessNode = {
+        minX: nodeX - 0.5 * processNodeViewWidth,
+        minY: nodeY - 0.5 * processNodeViewHeight,
+        maxX: nodeX + 0.5 * processNodeViewWidth,
+        maxY: nodeY + 0.5 * processNodeViewHeight,
+        position: transformedPosition,
+        entity: processEvent,
+        type: 'processNode',
+      };
+      processesToIndex.push(indexedEvent);
+    }
+    for (const edgeLineSegment of edgeLineSegments) {
+      const transformedSegment: Vector2[] = [];
+      for (const point of edgeLineSegment) {
+        transformedSegment.push(applyMatrix3(point, dataSelectors.isometricTransformMatrix));
+      }
+      const [[x1, y1], [x2, y2]] = transformedSegment;
+      const indexedLineSegment: IndexedEdgeLineSegment = {
+        minX: Math.min(x1, x2),
+        minY: Math.min(y1, y2),
+        maxX: Math.max(x1, x2),
+        maxY: Math.max(y1, y2),
+        entity: transformedSegment,
+        type: 'edgeLine',
+      };
+      edgeLineSegmentsToIndex.push(indexedLineSegment);
+    }
+    tree.load([...processesToIndex, ...edgeLineSegmentsToIndex]);
+    return tree;
+  }
+);
+
+const boundingBox = composeSelectors(cameraStateSelector, cameraSelectors.viewableBoundingBox);
+
+const currentBoundingBox = createSelector(
+  (state: ResolverState) => state,
+  function (state) {
+    return boundingBox(state)(Date.now());
+  }
+);
+
+export const visibleProcessNodePositionsAndEdgeLineSegments = createSelector(
+  spatiallyIndexedEntities,
+  currentBoundingBox,
+  function visibleProcessNodePositionsAndEdgeLineSegments(tree, { maximum, minimum }) {
+    const [minX, minY] = minimum;
+    const [maxX, maxY] = maximum;
+    const visibleEntities = tree.search({
+      minX,
+      minY,
+      maxX,
+      maxY,
+    });
+    const visibleProcessNodePositions = visibleEntities.filter(
```

Comment:
Why not return the the entire array and let the react code deal with it. in react

```
visibleEntities.map(entity => {
  if (entity.type === 'edge') {
    // components
  } else {
    // different components
  }
})
```

Reply:
Based on past PR feedback I've gotten, I think the indication here would be to keep it as it is:
https://github.com/elastic/kibana/pull/64121#discussion_r427609242